README.md

```diff
@@ -123,6 +122,43 @@ KOTLINC_RELEASE = {
 kotlin_repositories(compiler_release = KOTLINC_RELEASE)
 ```
 
+## Friends/Internal support
+
+The rules support kotlin `internal` visibility (usually accessible only to the compilation unit)
+by allowing a library or test to specify that it is friends with another library like so:
+
+```
+kt_jvm_library(
+    name = "foo",
+    srcs = glob(["*.kt"]),
+    friend = "//some/other:target",
+    # ...
+)
+```
+
+> Note: declaring friends of a kt_android_library requires adding `_kt` to the target, e.g.
+> `//some/other:target_kt`.  This is because the `kt_android_library` rule is a macro with a
+> `kt_jvm_library` under the hood, and the surfaced rule is an android rule which doesn't have
+> kotlin aware bazel providers. This will be fixed in the future.
+
+This grants access to `internal` members of `//some/other:target`. A library can only be friends
+with one other library, which must be a kotlin-aware target. It inherits the module name from that
+library. Only one friend can be declared, because of the assumptions made here about module
+membership.  Since friendship can be transitive (see below), this constrains the visibility so it
+does not become an arbitrarily growing lattice of trust, defeating the purpose.
+
+Very common use-cases for this are:
+
+  * tests and test-libraries depending on internals of the system under test.
+  * clusters of targets that represent one logical unit, with public, private,
+     fake, testing-utilities, configuration, or other targets that comprise the
+     whole unit.
+
+Friendship has limited transitivity. Consider projects `C`, `B`, and `A` with a dependency
```

Comment:
Why is friendship transitive? Can we make it strict, like regular Java dependencies? It's easier to loosen transitivity than to make it stricter down the road.

Reply:
So the current implementation is that you declare one friend. This is how the orginal friends field worked for the test rule. So the transitivity lets the tree of things associated with the logical "module" to be realized.   If we went with the "friend-graph" approach (see #211 proposal) then it would be reasonable to make friends not-transitive (declare what you consume) but just have the "we're in the same logical module" constraint.

The problem, of course, is the same as iwht strict deps. There are lots fo cases where it's pure boilerplate.  Like how we export jsr-330 along with both dagger and guice, because there are no cases where you would declare the latter two and not the first. Likewise, this could force tests to explicitly declare their testlib and their system under test, when the testlib for that package always is friends with the system under test. So it could get needlessly verbose.  Because there's no performance gain, like with classpath ordering and strict-deps, I don't think it's a huge deal either way.

Reply:
As @vRallev pointed out, according to the JetBrains folks, -Xfriend-paths is intended to be transitive, by design. We could make it strict, but that's constraining the compiler feature - though if we moved to the friends-graph approach it wouldn't precisely matter so much. But ultimately the flag is intended to be transitive, so I'd want to press for handling things the way Jetbrains intends their flag to be used.